Improve Oxide scanner API

RobinMalfait opened the pull request:
This PR updates the API for interacting with the Oxide API. Until now, we used the name `scanDir(…)` which is fine, but we do way more work right now.

We now have features such as:

1. Auto source detection (can be turned off, e.g.: `@tailwindcss/vite` doesn't need it)
2. Scan based on `@source`s found in CSS files
3. Do "incremental" rebuilds (which means that the `scanDir(…)` result was stateful).

To solve these issues, this PR introduces a new `Scanner` class where you can pass in the `detectSources` and `sources` options. E.g.:

```ts
let scanner = new Scanner({
  // Optional, omitting `detectSources` field disables automatic source detection
  detectSources: { base: __dirname }, 

  // List of glob entries to scan. These come from `@source` directives in CSS.
  sources: [
    { base: __dirname, pattern: "src/**/*.css" },
    // …
  ],
});
```

The scanner object has the following API:

```ts
export interface ChangedContent {
  /** File path to the changed file */
  file?: string
  /** Contents of the changed file */
  content?: string
  /** File extension */
  extension: string
}
export interface DetectSources {
  /** Base path to start scanning from */
  base: string
}
export interface GlobEntry {
  /** Base path of the glob */
  base: string
  /** Glob pattern */
  pattern: string
}
export interface ScannerOptions {
  /** Automatically detect sources in the base path */
  detectSources?: DetectSources
  /** Glob sources */
  sources?: Array<GlobEntry>
}
export declare class Scanner {
  constructor(opts: ScannerOptions)
  scan(): Array<string>
  scanFiles(input: Array<ChangedContent>): Array<string>
  get files(): Array<string>
  get globs(): Array<GlobEntry>
}
```

The `scanFiles(…)` method is used for incremental rebuilds. It takes the `ChangedContent` array for all the new/changes files. It returns whether we scanned any new candidates or not.

Note that the `scanner` object is stateful, this means that we don't have to track candidates in a `Set` anymore. We can just call `getCandidates()` when we need it.

This PR also removed some unused code that we had in the `scanDir(…)` function to allow for sequential or parallel `IO`, and sequential or parallel `Parsing`. We only used the same `IO` and `Parsing` strategies for all files, so I just got rid of it.

RobinMalfait committed:
remove top-level `scan_files`

The only place we used this in was in the UI tests. This is also the
only spot where we used the public `IO` and `Parsing` enums.

Note: you'll notice that I temporary used the `scanDir()` function in
the UI test which is a bit ugly, but because we currently don't pass in
any `sources` or `base` options, it's almost a no-op. But it does expose
the `scanFiles` function we use for incremental rebuilds.

This will be cleaned up in the next few commits

RobinMalfait committed:
remove `IO` and `Parallel`

At this point we only use `IO::Parallel | Parsing::Parallel`, and it's
only used on the Rust side. Let's remove all of it. We can re-introduce
it if we need it again.

RobinMalfait committed:
remove unused functions

We don't use the `xxx_sync` functions anymore, let's get rid of them and
re-introduce them when we need them.

RobinMalfait committed:
use `Into::into`

This now uses `Into::into` everywhere which makes the code consistent,
shorter and nicer because we don't need to know what the _actual_ type
is.

RobinMalfait committed:
store internal `GlobEntry`

We are converting between the private and public versions of
`GlobEntry`. But we never expose the `GlobEntry` objects anymore, we
just store it for later use.

This means that we can use the "private" `tailwindcss_oxide::GlobEntry`
directly, and we don't need to map between the different versions.

RobinMalfait committed:
remove unnecessary `.clone()`

RobinMalfait committed:
move `is_allowed_content_path` and `resolve_allowed_paths` to `scanner` folder

RobinMalfait committed:
move `resolve_files` and `resolve_globs` to `scanner/auto_content`

RobinMalfait committed:
move `lazy_static!` up

RobinMalfait committed:
remove unnecessary path conversion

RobinMalfait committed:
create dedicated `AutoContent` struct

RobinMalfait committed:
introduce internal `Scanner`

The `Scanner` object will contain the implementation for both auto
content detection and scanning incoming source globs.

RobinMalfait committed:
optimize auto content detected globs and incoming source globs

RobinMalfait committed:
tmp: introduce new scanner API

RobinMalfait committed:
rename `scan_dir` tests to more generic `scanner`

RobinMalfait committed:
create `Scanner` object

RobinMalfait committed:
use new `Scanner` class

RobinMalfait committed:
read candidates from the `scanner`

No need to track it in both JS and Rust land separately.

RobinMalfait committed:
re-add tracing

thecrypticace commented:
I'll give this a more thorough look tomorrow but I have a few comments on naming:

1. I'd drop `get` from the method names. They're nouns anyway so it's self explanatory and the scanner is effectively "read-only" from the outside so the prefixes feel extra unnecessary. Also could / should we turn these into real JS getters? e.g. `get candidates()`? If not that's fine just wondering if we want to (don't feel strongly about it so 🤷‍♂️).

- `getCandidates()` -> `candidates()`
- `getFiles()` -> `files()`
- `getGlobs()` -> `globs()`

2. `autoContent` feels like the wrong name now that we've settled on `@source`. So maybe:
- `detectSources` (i think i like this one the best)
- `discoverSources`
- `autoSource`
- `sourceDetection`
- `sourceDiscovery`

3. Is there a future where the `base` dir for automatic source / content detection would be useful elsewhere? Seems maybe reasonable that `base` could be a top-level option? Don't feel strongly about this — just thinking out loud.

Overall though I'm a fan of this API. Feels more designed and more open to future expansion than the current one.

philipp-spiess commented on crates/node/src/lib.rs:
```diff
@@ -77,67 +56,85 @@ impl From<GlobEntry> for tailwindcss_oxide::GlobEntry {
+  }
+
+  #[napi]
+  pub fn get_candidates(&self) -> Vec<String> {
+    self.scanner.get_candidates()
+  }
```
What is the behavior if a vec is returned here and exported to Node via napi? My current thinking is that this needs to copy the whole array every time we call getCandidates which might be quite a bit of I/O for large projects.

Maybe alternatively we could only return the new candidates (so moving some of the bookeeping into rust by keeping it in a set that we can discriminate between new and old).

I understand that it is annoying to collect the results in JS land in another data structure but I imagine that we would otherwise call `getCandidates()` quite often in dev mode so maybe this is a reasonable optimization?

RobinMalfait replied:
Yeah that's right. Thinking about it, I think we even track all known candidates in tailwind's core already. So this can definitely be simplified.

thecrypticace replied:
If we do that I think it needs a different name too — maybe something like "newCandidates" or w/e.

On the flip side — even with this stuff being copied — I'd see if we can measure the overhead to see if it's significant or not.

philipp-spiess reviewed (commented):
@thecrypticace 
> `autoContent` feels like the wrong name now that we've settled on @source. So maybe:

Yeah I agree, should be called something with source now 👍

thecrypticace commented on crates/oxide/src/scanner/allowed_paths.rs:
```diff
@@ -0,0 +1,59 @@
+use lazy_static::lazy_static;
+use std::path::Path;
+
+lazy_static! {
```
fyi now that we're on Rust 1.80 I think `lazy_static` can be replaced with `std::sync::LazyLock`

thecrypticace commented on crates/oxide/src/lib.rs:
```diff
@@ -65,24 +74,89 @@ pub fn clear_cache() {
+            ..Default::default()
+        };
+
+        scanner.scan();
```
It feels super weird to me that `::new()` has the side effect of touching the filesystem. I think I'd prefer that the users for `Scanner` call `scan` instead of relying on `::new()` basically not returning until its done.

RobinMalfait replied:
Yep I agree, reasons I still did it is that when using it, we basically always did this:

```ts
let scanner = new Scanner(…);
scanner.scan();
```

I think in a perfect world you can do something like this:
```ts
let scanner = new Scanner(…);

// Setup watchers, so that they are already "listening" for changes before we perform the first scan. Prevents a potential race condition where a file changed between setting up the scanner and creating the watchers.
await setupWatchers();

// Perform initial scan
scanner.scan();
```

However, in this case, in order to setup the watchers, we need to resolve the globs (which touches the file system).

thecrypticace commented on crates/oxide/src/lib.rs:
```diff
@@ -65,24 +74,89 @@ pub fn clear_cache() {
-            let globs = resolve_globs(base, dirs);
+impl Scanner {
+    pub fn new(auto_content: Option<AutoContent>, sources: Option<Vec<GlobEntry>>) -> Self {
+        init_tracing();
```
also thinking maybe this should be not in ::new() either but can't think of a good place to put it. Maybe it should be in `scan`?

thecrypticace commented:
Since `Scanner` is stateful I think the cache should live in / be owned by the `Scanner` object. Removes our need for a `clearCache()` top-level fn which always felt like a hack.

RobinMalfait committed:
inline `Cache` inside of the `Scanner`

This has some benefits:

1. The cache is properly scopes. New instances of `Scanner` will get
   their own cache.
   - A small internal side effect is that we can remove the serial tests
     because everything will be properly scoped now.
2. We can track all candidates and newly found candidates since last
   time we requested the list of candidates separately. This way we have
   a small set of new candidates to deal with.

RobinMalfait committed:
remove serial tests

RobinMalfait committed:
remove `clear_cache` function

RobinMalfait committed:
ensure we get candidates from the previous scanner

In Vite we always setup a new Scanner, so we have to make sure that we
get the candidates from the last scanner.

In the future we will properly scope the scanner to the `css` so that
this doesn't happen. (Because the scanner will be re-used.)

RobinMalfait committed:
replace `lazy_static!` with native `LazyLock`

RobinMalfait committed:
drop `get` prefix in functions

RobinMalfait committed:
make `files`, `globs` and `candidates` getters

RobinMalfait committed:
improve APIs

- `let scanner = new Scanner()`
- `scanner.scan()` // Returns candidates
- `scanner.scanFiles([])` // Returns new candidates from pass changed
  content
- `scanner.files` // Getter for getting computed files
- `scanner.globs` // Getter for getting computed globs

RobinMalfait committed:
only find files once

This is a bit of a weird one, but hear me out.

The `scanner.scan()` files has multiple responsibilities:

1. Figure out files and globs from the file system based on auto content
   detection.
2. Figure out files and globs from the files system based on `@source`
   (GlobEntry) directives.
3. Actually scan those files to extract candidates from them.

This means that if you want to access `scanner.files` or `scanner.globs`
(to setup watchers, or register them with PostCSS) that you _have_ to
call `scanner.scan()` first.

Setting up a watcher (e.g.: inside of the CLI) is an async action. So it
_could_ be that a file changed in between the `scanner.scan()` and
`await setupWatchers()` which could be a race condition.

Another issue is that we call `let candidates = scanner.scan()` but we
only need it later, but moving it closer to where we need it breaks the
`scanner.files` and `scanner.globs` code...

So instead, we want to split up the responsibility of `scan()` to first
figure out all files/globs then scan the files for candidates.

If we do that, we have to introduce a new function like
`scanner.prepareFilesAndGlobs()` which feels silly because you always
have to do that before accessing `scanner.files` or `scanner.globs`. In
fact, you also have to do that before calling `scanner.scan()`.

Another solution is to make `scanner.files()` and `scanner.globs()`
functions. Whenever these are called we do the discovery of those files.
However, this now means that we do it `n` times instead of once.

To solve this, we can track some state internally to know whether we did
that work already or not.

This brings us to the next issue. If you, for some reason, don't even
need `scanner.files` or `scanner.globs` (e.g.: an initial build without
any watchers) then `scanner.scan()` won't be able to scan the files
because they are not discovered yet.

So instead, we did this:
1. Setup a scanner
2. When accessing `scanner.files`, `scanner.globs`, calling
   `scanner.scan()` or calling `scanner.scanFiles()` we will discover
   the files/globs first.

Internally we do track whether this work has been done before or not.
But it allows us to delay the discovery until the first time you
actually need it. It allows us to not call a separate function (that you
can forget to call).

It does make the getter on the Rust side side-effect-y but that should
still be safe since JavaScript is single threaded.

thecrypticace committed:
Only calculate known / bundled extensions once

thecrypticace committed:
Merge branch 'next' into fix/improve-scan-dir

thecrypticace committed:
Fix UI tests

thecrypticace commented on packages/tailwindcss/tests/ui.spec.ts:
```diff
@@ -317,6 +316,7 @@ async function render(page: Page, content: string) {
 
   await page.setContent(content)
 
+  let scanner = new Scanner({})
```
fyi @RobinMalfait I fixed the UI tests — scanner being stateful caused problems because scanFiles only returns new candidates now. I think using a new scanner every time here is probably a good idea regardless.

thecrypticace committed:
Tweak `scan_sources`

This canonicalizes the root paths before creating the glob entries. This should end up with less work done — though it was minimal before

thecrypticace commented:
Something else to think about:

Intellisense does *not* need candidates when scanning for thing using Oxide. Just a list of files and globs is sufficient. Scanning the files themselves would be unnecessary work.

RobinMalfait commented:
> Something else to think about:
> 
> 
> 
> Intellisense does *not* need candidates when scanning for things using Oxide. Just a list of files and globs is sufficient. Scanning the files themselves would be unnecessary work.

Thats should already be the case now. You just have to use `scanner.files` and `scanner.globs`. Whenever the first one is used, the files/globs will be calculated.

`scanner.scan()` should not be required anymore.

RobinMalfait committed:
run `cargo fmt`

RobinMalfait committed:
rename `AutoContent` to `DetectSources`
